internal/controller/buildpipeline/buildpipeline_adapter.go

```diff
@@ -650,3 +691,112 @@ func generateDetails(buildPLR *tektonv1.PipelineRun, integrationTestStatus intgt
+			a.logger.Info(fmt.Sprintf("The build pipelineRun %s/%s with pr group %s is not the latest for its component, skipped", pipelineRun.Namespace, pipelineRun.Name, prGroup))
+			continue
+		}
+		if metadata.HasAnnotation(&pipelineRun, tekton.SnapshotNameLabel) {
+			a.logger.Info(fmt.Sprintf("The build pipelineRun %s/%s with pr group %s has snapshot created, skipped", pipelineRun.Namespace, pipelineRun.Name, prGroup))
+			continue
+		}
+		componentsFromPipelineRun = append(componentsFromPipelineRun, pipelineRun.Labels[tekton.PipelineRunComponentLabel])
+	}
+	return componentsFromPipelineRun, nil
+}
+
+// isGroupSnapshotExpectedForBuildPLR return group context ITS if group snapshot is expected for the same pr group with the given build PLR
+func (a *Adapter) isGroupSnapshotExpectedForBuildPLR(pipelineRun *tektonv1.PipelineRun) (bool, error) {
+	prGroupHash, prGroup := gitops.GetPRGroup(pipelineRun)
+	componentsFromPipelineRun, err := a.getComponentFromLatestFlightBuildPLR(prGroup, prGroupHash)
+	if err != nil {
+		a.logger.Error(err, fmt.Sprintf("failed to get component from build pipelineruns for pr group %s", prGroup))
+		return false, err
+	}
+
+	componentsFromSnapshot, err := a.loader.GetComponentsFromSnapshotForPRGroup(a.context, a.client, pipelineRun.Namespace, prGroup, prGroupHash)
+	if err != nil {
+		a.logger.Error(err, fmt.Sprintf("failed to get component from component snapshot for pr group %s", prGroup))
+		return false, err
+	}
+
+	joinedComponents := gitops.MergeAndDeduplicate(componentsFromPipelineRun, componentsFromSnapshot)
+	if len(joinedComponents) < 2 {
+		a.logger.Info(fmt.Sprintf("The number %s of components affected by this PR group %s is less than 2, skipping group snapshot creation", joinedComponents, prGroup))
+		return false, nil
+	}
+
+	applicationComponents, err := a.loader.GetAllApplicationComponents(a.context, a.client, a.application)
+	if err != nil {
+		return false, err
+	}
+
+	// check if there is opened MR/PR for the components
+	var componentsWithOpenPRMR []string
+	for _, applicationComponent := range *applicationComponents {
+		applicationComponent := applicationComponent // G601
+
+		if slices.Contains(componentsFromPipelineRun, applicationComponent.Name) {
+			a.logger.Info(fmt.Sprintf("There is in progress build pipelineRun for component %s/%s, won't check its component snapshot's pull/merge request", a.pipelineRun.Namespace, applicationComponent.Name))
+			componentsWithOpenPRMR = append(componentsWithOpenPRMR, applicationComponent.Name)
+			continue
+		}
+
+		if slices.Contains(componentsFromSnapshot, applicationComponent.Name) {
+			a.logger.Info(fmt.Sprintf("There is component snapshot for component %s/%s, will check its component snapshot MR/PR status", a.pipelineRun.Namespace, applicationComponent.Name))
+			var foundSnapshotWithOpenedPR *applicationapiv1alpha1.Snapshot
+			foundSnapshotWithOpenedPR, err = a.findSnapshotWithOpenedPR(&applicationComponent, a.pipelineRun.Namespace, prGroupHash)
```

Comment:
It seems to me that only the `Name` of the applicationComponent is used in the `findSnapshotWithOpenedPR` function. If we modified the function to take the `Name` string, we could greatly simplify [this loop](https://github.com/konflux-ci/integration-service/pull/1049/files#diff-3540d1f95a378c2e6bea89135278f11b57b55c7c4e0a7ac892eb6dfd251b5b55R749) so it doesn't require fetching all applications components and looping through them, and we could simply loop through `componentsFromSnapshot` instead.

Reply:
We also wouldn't need to append the components from in-flight builds [here](https://github.com/konflux-ci/integration-service/pull/1049/files#diff-3540d1f95a378c2e6bea89135278f11b57b55c7c4e0a7ac892eb6dfd251b5b55R752-R756) and could simply check that list directly when making a decision if a group Snapshot will be created.
Basically something like this:

- If `componentsFromPipelineRun` is more than 1, return `true` immediately since we know that a group Snapshot will need to be created for at least those ones, so no need to check further
- Otherwise loop through the `componentsFromSnapshot` and check if they have open PRs.
- At the end, if there are more than 1 `componentsFromSnapshot` that have open PRs, return `true`

Reply:
Updated. Thanks!